core/src/main/scala/cats/data/HashMap.scala

```diff
@@ -0,0 +1,1211 @@
+
+    final override def next(): (K, V) = {
+      if (!hasNext) throw new NoSuchElementException
+      val value = currentNode.getMapping(currentValuesIndex)
+      currentValuesIndex += 1
+      value
+    }
+  }
+}
+
+sealed abstract private[data] class HashMapInstances extends HashMapInstances1 {
+  implicit def catsDataUnorderedTraverseForHashMap[K: Hash]: UnorderedTraverse[HashMap[K, *]] =
+    new UnorderedTraverse[HashMap[K, *]] {
+      override def nonEmpty[A](fa: HashMap[K, A]): Boolean = fa.nonEmpty
+
+      override def isEmpty[A](fa: HashMap[K, A]): Boolean = fa.isEmpty
+
+      override def size[A](fa: HashMap[K, A]): Long = fa.size.toLong
+
+      override def contains_[A](fa: HashMap[K, A], v: A)(implicit ev: Eq[A]): Boolean =
+        fa.iterator.exists { case (_, value) => ev.eqv(v, value) }
+
+      override def exists[A](fa: HashMap[K, A])(p: A => Boolean): Boolean =
+        fa.iterator.exists { case (_, value) => p(value) }
+
+      override def forall[A](fa: HashMap[K, A])(p: A => Boolean): Boolean =
+        fa.iterator.forall { case (_, value) => p(value) }
+
+      override def count[A](fa: HashMap[K, A])(p: A => Boolean): Long =
+        fa.iterator.foldLeft(0L) { case (c, (_, value)) => if (p(value)) c + 1L else c }
+
+      def unorderedFoldMap[U, V](hm: HashMap[K, U])(f: U => V)(implicit V: CommutativeMonoid[V]): V =
+        V.combineAll(hm.iterator.map { case (_, u) => f(u) })
+
+      def unorderedTraverse[G[_], U, V](hashMap: HashMap[K, U])(f: U => G[V])(implicit
+        G: CommutativeApplicative[G]
+      ): G[HashMap[K, V]] = {
+        val emptyHm: Eval[G[HashMap[K, V]]] =
+          Always(G.pure(HashMap.empty[K, V]))
+
+        val gHashMap = Foldable
+          .iterateRight(() => hashMap.iterator, emptyHm) { case ((k, u), hm) =>
+            G.map2Eval(f(u), hm) { (v, map) =>
+              map.updated(k, v)
+            }
+          }
+
+        gHashMap.value
+      }
+    }
+
+  implicit def catsDataCommutativeMonoidForHashMap[K: Hash, V: CommutativeSemigroup]: CommutativeMonoid[HashMap[K, V]] =
+    new HashMapMonoid[K, V] with CommutativeMonoid[HashMap[K, V]]
+
+  implicit def catsDataShowForHashMap[K: Show, V: Show]: Show[HashMap[K, V]] =
+    Show.show[HashMap[K, V]](_.show)
+
+  implicit def catsDataHashForHashMap[K, V: Hash]: Hash[HashMap[K, V]] =
+    new Hash[HashMap[K, V]] {
+      def hash(hm: HashMap[K, V]): Int = hm.hash
+      def eqv(x: HashMap[K, V], y: HashMap[K, V]): Boolean = x === y
+    }
+}
+
+sealed abstract private[data] class HashMapInstances1 {
+  implicit def catsDataMonoidForHashMap[K: Hash, V: Semigroup]: Monoid[HashMap[K, V]] =
+    new HashMapMonoid[K, V]
+}
+
+class HashMapMonoid[K: Hash, V](implicit V: Semigroup[V]) extends Monoid[HashMap[K, V]] {
+
+  def empty: HashMap[K, V] = HashMap.empty[K, V]
+
+  def combine(xs: HashMap[K, V], ys: HashMap[K, V]): HashMap[K, V] = {
+    if (xs.size <= ys.size) {
```

Comment:
this should be the same code as `concat` with `V: Semigroup` being `(x, y) => y`

Can we inline this with `concat` to get the benefit of only iterating the smaller item (which we don't do in concat).

Reply:
It feels like we could extract this pattern as something like:

```scala
  def combineWith(xs: HashMap[K, V], ys: HashMap[K, V])(f: (HashMap.Node[K, V], (K, V)) => HashMap.Node[K, V]): HashMap[K, V] = {
    val newRootNode = if (xs.size <= ys.size) {
      xs.iterator.foldLeft(ys.rootNode) { case (node, kv @ (k, x)) =>
        f(node, kv)
      }
    } else {
      ys.iterator.foldLeft(xs.rootNode) { case (node, kv @ (k, y)) =>
        f(node, kv)
      }
    }

    if (newRootNode eq xs.rootNode)
      xs
    else if (newRootNode eq ys.rootNode)
      ys
    else
      new HashMap(newRootNode)
  }
```
No idea whether it's small enough to get JITted or how much the extra branch would hurt performance though

Reply:
There's one other subtlety to this change - it breaks the behaviour of `concat` a bit because it changes which value is kept when the key exists in both maps.
I think that the scala collections work around this by adding a boolean flag to `updated` which determines whether the existing value should be replaced.

Reply:
I have added a boolean parameter `replaceExisting` to `Node#updated`. We can use this to keep the existing key when iterating over the "rightmost" HashMap in the various combining operations.